decisions/0007-remix-on-react-router-6-4-0.md

```diff
@@ -0,0 +1,166 @@
+
+#### Notes
+
+- This can't use `process.env` since the code we're changing is runtime agnostic. We'll go with a local hardcoded variable in `server.ts` for now to avoid runtime-specific ENV variable concerns.
+  - Unit and integration tests may need to have their own copies of this variable as well to remain passing. For example, we have unit tests that assert that a loader is called once for a given route - but when this flag is enabled, that loader will be called twice so we can set up a conditional assertion based on the flag.
+- The `remixContext` sent through `entry.server.ts` will be altered in shape. We consider this an opaque API so not a breaking change.
+
+#### Implementation approach
+
+1. Use `createHierarchicalRoutes` to build RR `DataRouteObject` instances
+   1. See `createStaticHandlerDataRoutes` in the `brophdawg11/rrr` branch
+2. Create a static handler per-request using `unstable_createStaticHandler`
+3. `handleResourceRequest`
+   1. This one should be _really_ simple since it should just send back the raw `Response` from `queryRoute`
+4. `handleDataRequest`
+   1. This is only slightly more complicated than resource routes, as it needs to handle serializing errors and processing redirects into 204 Responses for the client
+5. `handleDocumentRequest`
+   1. This is the big one. It simplifies down pretty far, but has the biggest surface area where some things don't quite match up
+   2. We need to map query "errors" to Remix's definition of error/catch and bubble them upwards accordingly.
+      1. For example, in a URL like `/a/b/c`, if C exports a a `CatchBoundary` but not an `ErrorBoundary`, then it'll be represented in the `DataRouteObject` with `hasErrorBoundary=true` since the `@remix-run/router` doesn't distinguish
+      2. If C's loader throws an error, the router will "catch" that at C's `errorElement`, but we then need to re-bubble that upwards to the nearest `ErrorBoundary`
+      3. See `differentiateCatchVersusErrorBoundaries` in the `brophdawg11/rrr` branch
+   3. New `RemixContext`
+      1. `manifest`, `routeModules`, `staticHandlerContext`, `serverHandoffString`
+      2. Create this alongside `EntryContext` assert the values match
+   4. If we catch an error during render, we'll have tracked the boundaries on `staticHandlerContext` and can use `getStaticContextFromError` to get a new context for the second pass (note the need to re-call `differentiateCatchVersusErrorBoundaries`)
+
+### Do the UI rendering layer second
+
+The rendering layer in `@remix-run/react` is a bit more of a whole-sale replacement and comes with backwards-compatibility concerns, so it makes sense to do second. However, we can still do this iteratively, we just can't deploy iteratively since the SSR and client HTML need to stay synced (and associated hooks need to ead crom the same contexts). First, we can focus on getting the SSR document rendered properly without `<Scripts/>`. Then second we'll add in client-side hydration.
```

Comment:
```suggestion
The rendering layer in `@remix-run/react` is a bit more of a whole-sale replacement and comes with backwards-compatibility concerns, so it makes sense to do second. However, we can still do this iteratively, we just can't deploy iteratively since the SSR and client HTML need to stay synced (and associated hooks need to read from the same contexts). First, we can focus on getting the SSR document rendered properly without `<Scripts/>`. Then second we'll add in client-side hydration.
```